### DISCUSSION TITLE
cmake project file for application using Fast-DDS as dependency

### DISCUSSION THREAD

**Discussion by aldas** [eProsima/Fast-DDS#5278] [Ideas]
It would be nice if there would be somewhere example for cmake project file for C++ application using Fast-DDS as depencency with does not rely on system wide depencies (installed by apt).    This is not only my problem - for example here https://discourse.cmake.org/t/find-package-for-non-installer-third-party-module-library-package/8954 is discussion about similar/same thing.

I am developing DDS application. I do not use ROS2. I would like not to install depencies with apt as https://fast-dds.docs.eprosima.com/en/latest/installation/sources/sources_linux.html#dependencies describes `sudo apt install libasio-dev libtinyxml2-dev`  + foonathanmemory.

I have almost zero experience with Cmake and I managed to get something to build and compile, but I do not know if this a adequate solution

Project files structure:
```
CMakeLists.txt
main.cpp
```

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.20)
project(MyAppExample VERSION 1 LANGUAGES CXX)

set(EXECUTABLE_NAME "MyFastDDSDemo")

# versions for Fast-DDS and dependencies
set(FASTDDS_VERSION "v3.0.1")
set(FOONATHAN_MEMORY_VERSION "v1.3.1")


set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

# Define the installation directory
set(INSTALL_DIR "${CMAKE_CURRENT_BINARY_DIR}/install")

include(ExternalProject)
ExternalProject_Add(
        foonathanMemoryVendor
        GIT_REPOSITORY "https://github.com/eProsima/foonathan_memory_vendor.git"
        GIT_TAG ${FOONATHAN_MEMORY_VERSION}
        INSTALL_DIR ${INSTALL_DIR}/foonathan_memory_vendor
        CMAKE_ARGS
        -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
        -DBUILD_SHARED_LIBS=ON
)

ExternalProject_Add(FastDDS
        GIT_REPOSITORY https://github.com/eProsima/Fast-DDS.git
        GIT_TAG ${FASTDDS_VERSION}
        INSTALL_DIR ${INSTALL_DIR}/fastdds
        CMAKE_CACHE_ARGS
        -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
        -DCMAKE_PREFIX_PATH:PATH=${CMAKE_PREFIX_PATH};${INSTALL_DIR}
        -DBUILD_SHARED_LIBS:BOOL=ON
        -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
        -DCMAKE_TOOLCHAIN_FILE:PATH=${CMAKE_TOOLCHAIN_FILE}
        -DTHIRDPARTY:BOOL=ON  # install asio, tinyxml2 from Git
        -DSECURITY:BOOL=OFF
)

include_directories(
        ${INSTALL_DIR}/fastdds/include
        ${INSTALL_DIR}/foonathan_memory_vendor/include
)

add_executable(${EXECUTABLE_NAME} main.cpp)

add_dependencies(FastDDS foonathanMemoryVendor)
add_dependencies(${EXECUTABLE_NAME} FastDDS)

target_link_libraries(${EXECUTABLE_NAME}
        ${INSTALL_DIR}/fastdds/lib/libfastcdr.so
        ${INSTALL_DIR}/fastdds/lib/libfastdds.so
        ${INSTALL_DIR}/foonathan_memory_vendor/lib/libfoonathan_memory-0.7.3.so
)
```

main.cpp:
```cpp
#include <iostream>
#include <fastdds/dds/domain/DomainParticipant.hpp>
#include <fastdds/dds/domain/DomainParticipantFactory.hpp>

using namespace eprosima::fastdds::dds;


int main() {
    DomainParticipantFactory *factory = DomainParticipantFactory::get_instance();
    DomainParticipant *participant = factory->create_participant(0, PARTICIPANT_QOS_DEFAULT);

    if (participant == nullptr) {
        std::cerr << "Failed to create Domain Participant." << std::endl;
        return 1;
    }

    std::cout << "Domain Participant created successfully!" << std::endl;

    // Clean up
    factory->delete_participant(participant);
    return 0;
}

```

-------

p.s. why `foonathan_memory_vendor` needs to be added separately and not build like other thirdparty libs?

TLDR:  It would be nice if there would be some official bare minimum skeleton c++ project that you take and try to hack your things in it.

**Comment by aldas** [comment]
Dear reader, if you see potential improvements with that `CMakeLists.txt` please discuss it here.

**Comment by JesusPoderoso** [comment]
Hi @aldas, thanks for using Fast DDS.

> why foonathan_memory_vendor needs to be added separately and not build like other thirdparty libs?

Including the foonathan dependency as a thridparty is not part of our roadmap.

> TLDR: It would be nice if there would be some official bare minimum skeleton c++ project that you take and try to hack your things in it.

Have you tried the Fast DDS [examples](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/hello_world)?
There is also the example generation tool based on an idl with Fast DDS-Gen:

``` bash
fastddsgen -example CMake custom_data_type.idl
cmake . && cmake --build .
./custom_data_type publisher
```

**Reply by ploe24** [reply]
> Including the foonathan dependency as a thridparty is not part of our roadmap.

Hi @JesusPoderoso ,

I also have a similar issue.

My msvc17 (VS 22) CMake project uses FastDDS among other things as CMake submodule in its own dedicated thirdparty directory.
To configure and build recursive I use the `THIRDPARTY` option to dynamically build `ASIO`, `TinyXML` and `FastCDR` together/before FastDDS.
However `Foonathan_memory` / `Foonathan_memory_vendor` is a dependency which is not included but `FastDDS` depends on it as stated in the documentation.
So I also added it as submodule to my project... and now?!

When the FastDDS configure step reaches the point where it searches for `Foonathan`, it still fails... :(
The `foonathan_memory_vendor` documenation is quite poor compared to FastDDS.
I'm not sure what options or flags I need to set so that FastDDS in `MyProject/thirdparty/FastDDS` finds `foonathan_memory` in `MyProject/thirdparty/Foonathan`?! Both paths are added to CMake but it still reports

> CMake Error at thirdparty/Fast-DDS/CMakeLists.txt:261 (find_package):
  By not providing "Findfoonathan_memory.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "foonathan_memory", but CMake did not find one.
  
  
  I already specified `foonathan_memory_DIR` as suggested by CMake to no success :/
  
  The "manual" build instructions from source [here ](https://fast-dds.docs.eprosima.com/en/latest/installation/sources/sources_windows.html#local-installation) show that Foonathan was built in a FastDDS subdir... Is this needed?
  How to make it work when having CMake submodules?
  
  (Win 10 x64, CMake 3.31, msvc17 / VS 2022, FastDDS 3.1.0)
  
  Thanks for any help
  
  
  Edit:
  
  This topic 
  
  * https://discourse.cmake.org/t/find-package-for-non-installer-third-party-module-library-package/8954
  
  from the CMake discussion also describes my struggle.
  
The problem seems to be that `foonathan_memory_vendor` adds the "real" foonathan_memory by using `ExternalProject_add(...)` which I can't get to work when running the eProsima Foonathan vendor as submodule.
Is it better in this case to use `foonathan_memory` from GitHub directly?!

**Reply by juanlofer-eprosima** [reply]
Hi @ploe24 ,

I am not sure if it would be possible to build Fast-DDS in one cmake step with the submodules structure you are describing. If possible, I would try to first build foonathan's memory vendor package with `-DCMAKE_INSTALL_PREFIX` pointing to a local build folder, as well as passing `-DBUILD_SHARED_LIBS=ON`. Then, proceed to build Fast-DDS using `-DCMAKE_PREFIX_PATH` pointing at the path where foonathan was built (previous `CMAKE_INSTALL_PREFIX`).

**Reply by MiguelCompany** [reply]
The guys from Realsense have a way of building Fast DDS that may be of interest: https://github.com/IntelRealSense/librealsense/blob/master/CMakeLists.txt#L62-L75

**Reply by ploe24** [reply]
@juanlofer-eprosima @MiguelCompany 

Thanks for your reply, I will look into that.

The issue is/was that it works when I split the CMake config after Foonathan, open VS, build the "external" target that was added by the foonathan vendor and then continue with configuring/building FastDDS.
But I wanted to do this all in one step as shown in the CMakeList linked by @MiguelCompany :)
So RealSense makes use of `FetchContent` in their `foonathan.cmake` to pull it before FastDDS is looking for dependencies...
I think I'm also gonna try it this way. Looks promising.

Thanks again
